hack/check-state.sh

```diff
@@ -0,0 +1,132 @@
+
+
+        if [ "x$?" == "x0" ]; then
+            break
+        fi
+        if [ "$retry_count_hco" -ge "$MAX_GET_HCO_RETRY" ]; then
+            echo "can't get cluster state. commmand failed repeatedly. $CMD "
+            exit 1
+        fi
+        sleep 5
+        ((retry_count_hco=retry_count_hco+1))
+    done
+}
+
+function get_reconcile_json() {
+    RECONCILE_COMPLETED_JSON=$(echo "$HCO_JSON" | sed -e 's/.*ReconcileComplete: "\([^"]*\)" "\([^"]*\)" "\([^"]*\)".*/Status: \1 Reason: \2 Message: \3/')   
+    RECONCILE_COMPLETED=$(echo "$HCO_JSON" | sed -e 's/.*ReconcileComplete: "\([^"]*\)".*/\1/')   
+    if [ "x$RECONCILE_COMPLETED" != 'xTrue' ] && [ "x$RECONCILE_COMPLETED" != 'xFalse' ]; then
+        echo "Error: ReconcileCompleted not valid: $RECONCILE_COMPLETED_JSON"
+        exit 1
+    fi
+}
+
+function get_application_available_json() {
+    APPLICATION_AVAILABLE_JSON=$(echo "$HCO_JSON" | sed -e 's/.*ApplicationAvailable: "\([^"]*\)" "\([^"]*\)" "\([^"]*\)".*/Status: \1 Reason: \2 Message: \3/')   
+    APPLICATION_AVAILABLE=$(echo "$HCO_JSON" | sed -e 's/.*ApplicationAvailable: "\([^"]*\)".*/\1/')   
+    if [ "x$APPLICATION_AVAILABLE" != 'xTrue' ] && [ "x$APPLICATION_AVAILABLE" != 'xFalse' ]; then
+        echo "Error: ApplicationAvailable not valid: $APPLICATION_AVAILABLE_JSON"
+        exit 1
+    fi
+}
+
+function get_operation_progressing_json() {
+    OPERATION_PROGRESSING_JSON=$(echo "$HCO_JSON" | sed -e 's/.*ApplicationProgressing: "\([^"]*\)" "\([^"]*\)" "\([^"]*\)".*/Status: \1 Reason: \2 Message: \3/')   
+    OPERATION_PROGRESSING=$(echo "$HCO_JSON" | sed -e 's/.*ApplicationProgressing: "\([^"]*\)".*/\1/')   
+     
+    if [ "x$OPERATION_PROGRESSING" != 'xTrue' ] && [ "x$OPERATION_PROGRESSING" != 'xFalse' ]; then
+        echo "Error: OperationProgressing not valid: $OPERATION_PROGRESSING_JSON"
+        exit 1
+    fi
+}
+
+function get_application_degraded_json() {
+    APPLICATION_DEGRADED_JSON=$(echo "$HCO_JSON" | sed -e 's/.*ApplicationDegraded: "\([^"]*\)" "\([^"]*\)" "\([^"]*\)".*/Status: \1 Reason: \2 Message: \3/')   
+    APPLICATION_DEGRADED=$(echo "$HCO_JSON" | sed -e 's/.*ApplicationDegraded: "\([^"]*\)".*/\1/')   
+  
+    if [ "x$APPLICATION_DEGRADED" != 'xTrue' ] && [ "x$APPLICATION_DEGRADED" != 'xFalse' ]; then
+        echo "Error: ApplicationDegraded not valid: $APPLICATION_DEGRADED_JSON"
+        exit 1
+    fi
+}
+
+function wait_until_cluster_up() {
+
+   retry_count=0
+   echo "Waiting for cluster to start. This check can take a long time..."
+   
+   while [ "$retry_count" -lt "$MAX_APP_AVAILABLE_CHECKS" ]; do
+     get_hco_json
+    
+     get_reconcile_json
+     if [ "x$RECONCILE_COMPLETED" == 'xTrue' ]; then
+
+        get_application_available_json
+        get_application_degraded_json
+        
+        if [ "x$APPLICATION_AVAILABLE" == 'xTrue' ] && [ "x$APPLICATION_DEGRADED" == 'xFalse' ]; then
```

Comment:
I'm curious why "OPERATION_PROGRESSING" isn't checked here for false? Isn't there a case where the application is functioning but is in the middle of upgrading or healing?

Reply:
OPEATION_PROGRESSING is checked in line #87, this way I can put out a detailed print out in line #110 that we are still progressing.

Reply:
Just a question: what should happen if after the global timeout we have APPLICATION_AVAILABLE=True && APPLICATION_DEGRADED=False && OPERATION_PROGRESSING=True ? Is that considered a failure or success of the upgrade?

Reply:
I misunderstood that OPERATION_PROGRESSING==True while APPLICATION_AVAILABLE==True && APPLICATION_DEGRADED==False would mean that the upgrade finished successfully. Ok, will fix that.

Reply:
fixed in commit 9719329237c6e0adc65fc2f1bc34dc152926dcbb